test-common/src/main/scala/org/scalajs/testcommon/RPCCore.scala

```diff
@@ -0,0 +1,187 @@
+            send(makeReply(callId, repl))
+          } { bep =>
+            val ep: bep.endpoint.type = bep.endpoint
+            import ep._
+
+            import scala.concurrent.ExecutionContext.Implicits.global
+
+            futureFromTry(Try(deserialize[ep.Req](in)))
+              .flatMap(bep.exec)
+              .onComplete(repl => send(makeReply(callId, repl)))
+          }
+      }
+    }
+  }
+
+  /** Subclass needs to implement message sending. */
+  protected def send(msg: String): Unit
+
+  /** Used to make an actual call to the other end. */
+  final def call(ep: RPCEndpoint)(req: ep.Req): Future[ep.Resp] = {
+    import ep._
+
+    // Reserve an id for this call.
+    val id = nextId.incrementAndGet()
+
+    // Prepare message.
+    val msg = makeMsg(opCode, id, req)
+
+    // Register pending call.
+    val promise = Promise[Resp]
+    val oldCall = pending.put(id, PendingCall(promise))
+
+    if (oldCall != null) {
+      throw new AssertionError("Ran out of call ids!")
+    }
+
+    // Actually send message.
+    send(msg)
+
+    promise.future
+  }
+
+  /** Attaches the given method to the given (local) endpoint. */
+  final def attach(ep: RPCEndpoint)(ex: ep.Req => ep.Resp): Unit = {
+    attachAsync(ep)(x => futureFromTry(Try(ex(x))))
+  }
+
+  /** Attaches the given method to the given (local) endpoint. */
+  final def attachAsync(ep: RPCEndpoint)(ex: ep.Req => Future[ep.Resp]): Unit = {
+    require(ep.opCode != REPLY_OK && ep.opCode != REPLY_ERR, "Reserved op code")
+
+    val bound = new BoundEndpoint {
+      val endpoint: ep.type = ep
+      val exec = ex
+    }
+
+    val old = endpoints.put(ep.opCode, bound)
+    require(old == null, "Endpoint already was attached.")
+  }
+
+  final def detach(ep: RPCEndpoint): Unit = {
+    val old = endpoints.remove(ep.opCode)
+    require(old != null, "Endpoint was not attached.")
+  }
+
+  /** Close the communication channel. */
+  def close(): Unit = {
+    for {
+      (callId, p) <- pending.asScala
+      failing <- Option(pending.remove(callId))
+    } {
+      failing.promise.failure(new IOException("Channel got closed"))
+    }
+  }
+
+  private def makeReply[T : Serializer](id: Long, result: Try[T]): String = {
+    result.map(makeMsg(REPLY_OK, id, _)) match {
+      case Success(m) => m
+      case Failure(t) => makeMsg(REPLY_ERR, id, t)
+    }
+  }
+
+  private def makeMsg[T : Serializer](opCode: OpCode, id: Long,
+      payload: T): String = {
+    Serializer.withOutputStream { out =>
+      out.writeByte(opCode)
+      out.writeLong(id)
+      serialize(payload, out)
+    }
+  }
+
+  /** Same as Future.fromTry(x) but works in 2.10 */
+  private def futureFromTry[T](x: Try[T]): Future[T] = {
+    val promise = Promise[T]
+    promise.complete(x)
+    promise.future
+  }
+}
+
+private[scalajs] object RPCCore {
+  type OpCode = Byte
+
+  /** Exception thrown if a remote invocation fails. */
+  class RPCException(c: Throwable) extends Exception(null, c)
```

Comment:
Not statically, no. Yes, you can call `getClass.getName` and get `RPCException` back, but no publicly exposed type contains `RPCException`. So I don't see how this is a problem (anymore than returning an instance of a private type for `Object`).

I specifically do not want anyone to be able to match on this. It is only here for debugging purposes (but for this it's important to retain the stack trace of the RPC client). An alternative would be to throw just an `Exception` but that's also not very nice :S